contracts/hyfi/contract/src/model.rs

```diff
@@ -0,0 +1,174 @@
+pub struct Balances {
+    accounts: Vec<Account>,
+    pub collected_fee: u64,
+}
+impl Default for Balances {
+    fn default() -> Self {
+        Balances {
+            accounts: vec![
+                Account::new("faucet".into(), 1_000_000),
+                Account::new("test".into(), 1_000),
+            ],
+            collected_fee: 0,
+        }
+    }
+}
+impl TryFrom<sdk::StateDigest> for Balances {
+    type Error = Error;
+
+    fn try_from(state: sdk::StateDigest) -> Result<Self, Self::Error> {
+        let (balances, _) = bincode::decode_from_slice(&state.0, bincode::config::standard())
+            .map_err(|_| anyhow::anyhow!("Could not decode start height"))?;
+        Ok(balances)
+    }
+}
+
+impl sdk::Digestable for Balances {
+    fn as_digest(&self) -> sdk::StateDigest {
+        sdk::StateDigest(
+            bincode::encode_to_vec(self, bincode::config::standard())
+                .expect("Failed to encode Balances"),
+        )
+    }
+}
+
+impl Balances {
+    pub fn add_account(&mut self, account: Account) {
+        self.accounts.push(account);
+    }
+
+    pub fn get_account_index(&self, name: &str) -> Option<usize> {
+        self.accounts.iter().position(|acc| acc.name == name)
+    }
+
+    pub fn get_balance(&self, name: &str) -> Result<u64, Error> {
+        match self.get_account_index(name) {
+            Some(index) => Ok(self.accounts[index].balance),
+            None => bail!("Account '{}' does not exist", name),
+        }
+    }
+
+    pub fn send(&mut self, from: &str, to: &str, amount: u64) -> Result<(), Error> {
+        let from_index = match self.get_account_index(from) {
+            Some(index) => index,
+            None => bail!("Account '{}' does not exist", from),
+        };
+
+        self.accounts[from_index].send(amount)?;
+
+        let to_index = match self.get_account_index(to) {
+            Some(index) => index,
+            None => {
+                self.add_account(Account::new(to.to_owned(), 0));
+                self.accounts.len() - 1
+            }
+        };
+
+        self.accounts[to_index].receive(amount);
+        Ok(())
+    }
+
+    pub fn mint(&mut self, to: &str, amount: u64) -> Result<(), Error> {
+        let faucet_account_index = match self.get_account_index("faucet") {
+            Some(index) => index,
+            None => bail!("Account faucet does not exist. This should not happen"),
+        };
+
+        self.accounts[faucet_account_index].send(amount)?;
+
+        let to_index = match self.get_account_index(to) {
+            Some(index) => index,
+            None => {
+                self.add_account(Account::new(to.to_owned(), 0));
+                self.accounts.len() - 1
+            }
+        };
+        self.accounts[to_index].receive(amount);
+        Ok(())
+    }
+
+    pub fn pay_fees(&mut self, from: &str, amount: u64) -> Result<(), Error> {
+        let from_index = match self.get_account_index(from) {
```

Comment:
for diverging statements, we could use Lancelot's tip:
`let Some(from_index) = self.get_account_index(from) else bail!("Account '{}' does not exist", from);`